### PR TITLE
Add support for the official splicing protocol

### PR DESCRIPTION
**Pull request by t-bast**
We replace our experimental version of `splice_init`, `splice_ack` and `splice_locked` by their official version (see https://github.com/lightning/bolts/pull/1160). If our peer is using the experimental feature bit, we convert our outgoing messages to use the experimental encoding and incoming messages to the official messages.

We also change the TLV fields added to `tx_add_input`, `tx_signatures` and `splice_locked` to match the spec version. We always write both the official and experimental TLV to updated nodes (because the experimental one is odd and will be ignored) but we drop the official TLV if our peer is using the experimental feature, because they won't understand the even TLV field.

We do the same thing for the `commit_sig` TLV. For peers who support the official splicing version, we insert the `start_batch` message before the batch of `commit_sig` messages as specified in https://github.com/lightning/bolts/pull/1160.

We also support both the legacy and official reconnection logic (TLV fields in `channel_reestablish`). It is a bit painful because both protocols use TLV 1, but for different purposes, so we need to do it in a somewhat hacky way at the codec level.

This ensures that we're able to create splice transactions, even with non-upgraded nodes (full backwards-compatibility). It would be best to make sure Phoenix users upgrade to a version that uses the official protocol though, to allow cleaning up the codebase and guarantee that we haven't missed an edge case.

E2E tests:

- no issues making payments and reconnecting to lnd 0.20.1
- no issues making payments and reconnecting to cln 25.12.1
- no issues making payments and reconnecting to ldk with https://github.com/jkczyz/ldk-server/tree/2026-03-splicing-rbf (no issues with RBF either)

### CODE REVIEW

**Comment by remyers** [comment]
This is an observation that occurred to me while reviewing this PR that might be relevant for a future PR ..

There will be times when using `CMD_BUMP_FUNDING_FEE` instead of `CMD_SPLICE` would cost less on-chain and confirm faster if you want to perform an additional splice while a pending splice is unconfirmed. The total cost to create a new funding transaction with `CMD_SPLICE` will often be higher than making the same changes with `CMD_BUMP_FUNDING_FEE` at a slightly higher fee rate. If you are blocked from making a new splice because a pending splice has been fee bumped, you could also make a new fee bump to splice in/out value rather than waiting for the previously rbf'd splice to confirm.

This seems to be technically possible in the protocol because CMD_BUMP_FUNDING_FEE uses the interactive tx protocol and can change the funding contribution with the `funding_output_contribution` TLV. Figuring out if a splice or rbf-splice is cheaper will also need to consider the additional fees paid to bump our counter party's inputs and outputs.

Perhaps the biggest downside I can see (beside complexity) is that if your pending splice or rbf confirms before a rbf-splice, then you would need to renegotiate it as a new splice.

Am I missing something here wrt fee savings? Do you think it would be worth considering this situation now rather than later?

**Comment by t-bast** [comment]
> Perhaps the biggest downside I can see (beside complexity) is that if your pending splice or rbf confirms before a rbf-splice, then you would need to renegotiate it as a new splice.

Yes, the main reason we're not providing this yet (even though this is in theory possible) is to manage complexity. It can get really complex very quickly once you start going down that road, because if each splice "part" maps to a specific action (an on-the-fly funding), it can be a huge mess to reconcile which parts where actually done and which parts need to be replayed if a previous RBF attempt confirms.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala [hunk 1]
```diff
@@ -885,68 +918,73 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
...
             stay() using d.copy(spliceStatus = SpliceStatus.SpliceAborted) sending TxAbort(d.channelId, InvalidSpliceNotQuiescent(d.channelId).getMessage)
           } else if (msg.feerate < nodeParams.currentFeerates.minimum) {
             log.info("rejecting splice request: feerate too low")
             stay() using d.copy(spliceStatus = SpliceStatus.SpliceAborted) sending TxAbort(d.channelId, InvalidSpliceRequest(d.channelId).getMessage)
+          } else if (d.commitments.active.count(_.fundingTxIndex == d.commitments.latest.fundingTxIndex) > 1) {
+            val previousTxs = d.commitments.active.filter(_.fundingTxIndex == d.commitments.latest.fundingTxIndex).map(_.fundingTxId)
+            log.info("rejecting splice request: the previous splice has unconfirmed rbf attempts ({})", previousTxs.mkString(", "))
+            stay() using d.copy(spliceStatus = SpliceStatus.SpliceAborted) sending TxAbort(d.channelId, InvalidSpliceWithUnconfirmedRbf(d.channelId, previousTxs).getMessage)
           } else {
             log.info(s"accepting splice with remote.in.amount=${msg.fundingContribution} remote.in.push=${msg.pushAmount}")
```
Comment (remyers) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:989]: If 0-conf is activated, should we still only allow a limited number of previous unconfirmed parent txs? and/or have some sort of rate limit on?  or perhaps it doesn't matter if we already trust our peer enough to activate 0-conf..

**Review by remyers — CHANGES_REQUESTED** [review]
Looks good - the refactor for RBF makes it easier to read/understand.

I found a few nits but main issues to take a look at are related to adding logic to reject an TxInitRbf or SpliceInit when the parent splice is unconfirmed.

File: eclair-core/src/main/scala/fr/acinq/eclair/wire/internal/channel/version4/ChannelCodecs4.scala [hunk 2]
```diff
@@ -658,7 +658,7 @@ private[channel] object ChannelCodecs4 {
         ("remotePushAmount" | millisatoshi) ::
         ("waitingSince" | blockHeight) ::
         ("lastChecked" | blockHeight) ::
-        ("rbfStatus" | rbfStatusCodec) ::
+        ("status" | dualFundingStatusCodec) ::
         ("deferred" | optional(bool8, lengthDelimited(channelReadyCodec)))).as[DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED]
```
[the remark this answers is not in the document]
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/wire/internal/channel/version4/ChannelCodecs4.scala:662]: Ah! I see, it only actually reads/writes 0x01 or 0x02 in dualFundingStatusCodec to indicate if the state is waiting for sigs or waiting for confirmation.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala [hunk 3]
```diff
@@ -885,68 +918,73 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
...
             log.info("rejecting splice request: channel not quiescent")
             stay() using d.copy(spliceStatus = SpliceStatus.SpliceAborted) sending TxAbort(d.channelId, InvalidSpliceNotQuiescent(d.channelId).getMessage)
           } else if (msg.feerate < nodeParams.currentFeerates.minimum) {
             log.info("rejecting splice request: feerate too low")
             stay() using d.copy(spliceStatus = SpliceStatus.SpliceAborted) sending TxAbort(d.channelId, InvalidSpliceRequest(d.channelId).getMessage)
+          } else if (d.commitments.active.count(_.fundingTxIndex == d.commitments.latest.fundingTxIndex) > 1) {
+            val previousTxs = d.commitments.active.filter(_.fundingTxIndex == d.commitments.latest.fundingTxIndex).map(_.fundingTxId)
+            log.info("rejecting splice request: the previous splice has unconfirmed rbf attempts ({})", previousTxs.mkString(", "))
+            stay() using d.copy(spliceStatus = SpliceStatus.SpliceAborted) sending TxAbort(d.channelId, InvalidSpliceWithUnconfirmedRbf(d.channelId, previousTxs).getMessage)
           } else {
```
[the remark this answers is not in the document]
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:991]: Looks good!

**Review by remyers — CHANGES_REQUESTED** [review]
Changes to tests to disallow sequences of unconfirmed splices looks good. I only had a few minor nits/questions.

File: eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalSplicesStateSpec.scala [hunk 4]
```diff
@@ -1373,93 +1478,86 @@ class NormalSplicesStateSpec extends TestKitBaseClass with FixtureAnyFunSuiteLik
...
+    val fundingTx2 = initiateSplice(f, spliceIn_opt = Some(SpliceIn(500_000 sat, pushAmount = 0 msat)))
+    checkWatchConfirmed(f, fundingTx2)
     alice2bob.expectNoMessage(100 millis)
     bob2alice.expectNoMessage(100 millis)
-    // we now have two unconfirmed splices
+
+    // From Alice's point of view, we now have two unconfirmed splices.
 
     alice2bob.ignoreMsg { case _: ChannelUpdate => true }
     bob2alice.ignoreMsg { case _: ChannelUpdate => true }
```
[the remark this answers is not in the document]
Comment (remyers) [reply eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalSplicesStateSpec.scala:1498]: Ohh, I see. Its used before you do the disconnect/reconnect sequence.

File: eclair-core/src/test/scala/fr/acinq/eclair/io/OpenChannelInterceptorSpec.scala [hunk 5]
```diff
@@ -118,7 +118,7 @@ class OpenChannelInterceptorSpec extends ScalaTestWithActorTestKit(ConfigFactory
   test("add liquidity if on-the-fly funding is used", Tag(noPlugin)) { f =>
     import f._
 
-    val features = defaultFeatures.add(Features.SplicePrototype, FeatureSupport.Optional).add(Features.OnTheFlyFunding, FeatureSupport.Optional)
+    val features = defaultFeatures.add(Features.Splicing, FeatureSupport.Optional).add(Features.OnTheFlyFunding, FeatureSupport.Optional)
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/io/OpenChannelInterceptorSpec.scala]: To ease rebase on feature branches:
```suggestion
    val features = defaultFeatures
      .add(Features.Splicing, FeatureSupport.Optional)
      .add(Features.OnTheFlyFunding, FeatureSupport.Optional)
```

File: eclair-core/src/test/scala/fr/acinq/eclair/io/OpenChannelInterceptorSpec.scala [hunk 6]
```diff
@@ -210,7 +210,7 @@ class OpenChannelInterceptorSpec extends ScalaTestWithActorTestKit(ConfigFactory
   test("reject on-the-fly channel if another channel exists", Tag(noPlugin)) { f =>
     import f._
 
-    val features = defaultFeatures.add(Features.SplicePrototype, FeatureSupport.Optional).add(Features.OnTheFlyFunding, FeatureSupport.Optional)
+    val features = defaultFeatures.add(Features.Splicing, FeatureSupport.Optional).add(Features.OnTheFlyFunding, FeatureSupport.Optional)
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/io/OpenChannelInterceptorSpec.scala]:
```suggestion
    val features = defaultFeatures
      .add(Features.Splicing, FeatureSupport.Optional)
      .add(Features.OnTheFlyFunding, FeatureSupport.Optional)
```

File: eclair-core/src/test/scala/fr/acinq/eclair/io/OpenChannelInterceptorSpec.scala (hunk 5 above)
Comment (t-bast) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/OpenChannelInterceptorSpec.scala]: Done in https://github.com/ACINQ/eclair/pull/2887/commits/668becb06a7b7dc3e19a22b8cae00de097f1ee62

File: eclair-core/src/test/scala/fr/acinq/eclair/io/OpenChannelInterceptorSpec.scala (hunk 6 above)
Comment (t-bast) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/OpenChannelInterceptorSpec.scala]: Done in https://github.com/ACINQ/eclair/pull/2887/commits/668becb06a7b7dc3e19a22b8cae00de097f1ee62

File: eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LightningMessageTypes.scala [hunk 7]
```diff
@@ -143,14 +144,14 @@ case class TxSignatures(channelId: ByteVector32,
...
 }
 
 object TxSignatures {
   def apply(channelId: ByteVector32, tx: Transaction, witnesses: Seq[ScriptWitness], previousFundingSig_opt: Option[ChannelSpendSignature]): TxSignatures = {
     val tlvs: Set[TxSignaturesTlv] = previousFundingSig_opt match {
-      case Some(IndividualSignature(sig)) => Set(TxSignaturesTlv.PreviousFundingTxSig(sig))
+      case Some(IndividualSignature(sig)) => Set(TxSignaturesTlv.PreviousFundingTxSig(sig), TxSignaturesTlv.ExperimentalPreviousFundingTxSig(sig))
       case Some(partialSig: PartialSignatureWithNonce) => Set(TxSignaturesTlv.PreviousFundingTxPartialSig(partialSig))
       case None => Set.empty
     }
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LightningMessageTypes.scala]:
```scala
  def apply(channelId: ByteVector32, 
            tx: Transaction, 
            witnesses: Seq[ScriptWitness], 
            previousFundingSig_opt: Option[ChannelSpendSignature]): TxSignatures = {
    val tlvs: Set[TxSignaturesTlv] = Set(
      previousFundingSig_opt match {
        case Some(IndividualSignature(sig)) => Set(TxSignaturesTlv.PreviousFundingTxSig(sig), TxSignaturesTlv.ExperimentalPreviousFundingTxSig(sig))
        case Some(partialSig: PartialSignatureWithNonce) => Set(TxSignaturesTlv.PreviousFundingTxPartialSig(partialSig))
        case None => Set.empty
      },
    ).flatten
    TxSignatures(channelId, tx.txid, witnesses, TlvStream(tlvs))
  }
```

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LightningMessageTypes.scala]: I've done something slightly differently in https://github.com/ACINQ/eclair/pull/2887/commits/318ba314d5bc9824e031d07b772c84920ccfd6ff to flatten `Option`s instead of `Set`s, which should be even easier when rebasing.

**Review by pm47 — APPROVED** [review]
